### PR TITLE
Select Tool: Show block borders on `:hover` while using the Select tool

### PR DESCRIPTION
**Pull request by shaunandrews**
This PR adds a border to blocks while hovering with the Select tool active:

### CODE REVIEW

**Comment by mapk** [comment]
Shaun! This is looking pretty good. I've got some feedback after testing this out.

### 1. Border jumping
There's a weird action while hovering where the border jumps.

jump

### 2. Border contrast
The borders should adapt to allow proper color contrast levels depending on the background.

borders

### 3. Nested blocks
Should the user see every block they hover over? Or just the immediate child block border? Right now, as in the above gif, the user only would see the immediate child block's border.

Related: https://github.com/WordPress/gutenberg/issues/22383

**Comment by mtias** [comment]
Let's try to aim for just one block showing borders at a time.

**Comment by ZebulanStanphill** [comment]
Yeah, for simplicity, let's just show the hovered block's border. We can always change it later, and just getting this in will be a nice enhancement. I know there are some people who have complained that they can't see the borders of blocks, and this should at least help a little if not a lot.

I also think adding this hover interaction makes a lot of sense for a "select" mode. You're trying to select a block, so it makes sense that you would want to get an idea of which block you're about to select by seeing its borders. This will make Select mode a lot more useful for mouse users.

File: packages/block-editor/src/components/block-list/block-wrapper.js [hunk 1]
```diff
@@ -203,15 +241,18 @@ const BlockComponent = forwardRef(
 					className,
 					props.className,
 					wrapperProps && wrapperProps.className,
+					{ 'is-hovered': isBlockHovered },
 					! isAligned && 'wp-block'
 				) }
 				data-block={ clientId }
 				data-type={ name }
 				data-title={ blockTitle }
+				onMouseOver={ maybeHover }
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-list/block-wrapper.js:250]: Might be good to only add this for `isSelected`.

Comment (ellatrix) [reply packages/block-editor/src/components/block-list/block-wrapper.js:250]: Oh, never mind :) It should be added in all cases.

**Comment by shaunandrews** [comment]
With some help from @ellatrix I managed to get this working so that only a single block border appears as you move your cursor around. Here's how its looking now:

select-hover-border

--

> There's a weird action while hovering where the border jumps.

I've no clue what's causing this, but its less noticeable with the recent updates. I _think_ its something to do with the shape of the block's element changing for some unknown reason.

> The borders should adapt to allow proper color contrast levels depending on the background.

I'm not so sure. The current select border doesn't really adapt, and if you compare with tools like Figma or Sketch the border color is consistent regardless of colors on the canvas.

--

I do notice that with the Select tool, the borders don't display for blocks nested within a Group block. However, this seems to be related to the Select tool itself — if I enable the CSS styles for the Edit tool, I'm able to highlight nested blocks just fine.

**Comment by enriquesanchez** [comment]
Just took this for a spin and it feels good. I particularly like how the hover border is very thin and neutral.

Is there any reason why this hover effect is only available on the Select mode? I  feel this will be useful no matter what mode you're on.

**Comment by ZebulanStanphill** [comment]
> Is there any reason why this hover effect is only available on the Select mode? I feel this will be useful no matter what mode you're on.

We used to have block borders shown on hover all the time, but it was removed to make the editor look more like the front-end by default (only the selected block should have its appearance modified) and to reduce UI flashing as you move your cursor around. It was part of the whole G2 effort to lighten the UI.

I think it's a good idea to bring back hover borders in Select mode, though, since that mode is specifically designed for selecting blocks, and showing borders makes a lot of sense in that context.

**Comment by ZebulanStanphill** [comment]
I just tried out this branch. It seems to be working pretty well. The only thing I'm not sure about is the color of the hover border. It seems maybe a bit too light?

**Comment by shaunandrews** [comment]
The SCSS variable I used, `$light-gray-ui`, is listed as part of the recent G2 work in the [_colors.scss file](https://github.com/WordPress/gutenberg/blob/4fe937c2188e0677b03e6eb5d2c54292705f6c45/packages/base-styles/_colors.scss#L31) seemed like it works well to me. The name of the variable seems to match to intention here, but I'd be happy to change the hover color to something else.

**Comment by karmatosed** [comment]
For me the light-gray works as it is just enough to support the movement. I don't think it needs to be high contrast beyond that, but that could be explored as an iteration if proven to be a need. The color is less important in strength than the movement with an indicator.

File: packages/block-editor/src/components/block-list/style.scss [hunk 2]
```diff
@@ -161,8 +162,11 @@
 			}
 		}
 	}
-}
 
+	.is-navigate-mode & .block-editor-block-list__block.is-hovered:not(.is-selected)::after {
```
[the remark this answers is not in the document]
Comment (shaunandrews) [reply packages/block-editor/src/components/block-list/style.scss:166]: Yea, that'd probably work just fine I think.

[the remark this answers is not in the document]
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/style.scss:166]: Ok, I changed its so the class is only added in navigation mode. So this `.is-navigate-mode` can be removed here and above.

**Review by ellatrix — APPROVED** [review]
Looks good to me, if the colour doesn't need updating.

**Comment by ZebulanStanphill** [comment]
I'm not opposed to the current color, so feel free to merge if you want to!

File: packages/block-editor/src/components/block-list/style.scss (hunk 2 above)
[the remark this answers is not in the document]
Comment (ZebulanStanphill) [reply packages/block-editor/src/components/block-list/style.scss:165]: Perhaps the class name should be updated to reflect that it isn't applied every time the block is hovered, but only when it's hovered in navigation/select mode?

**Comment by noahtallen** [comment]
Is this supposed to still be working? :p

I am not seeing it work for any blocks locally currently. Not sure if bug or if it was removed on purpose!

**Comment by ZebulanStanphill** [comment]
@noahtallen It's working for me on the latest `master`.

image

**Comment by noahtallen** [comment]
Thanks! I think it is an issue with what I'm working on. The changes seemed unrelated which is why I started looking here :)

**Pull request merged**